### ISSUE TITLE
imageio ffmpeg reader: IndexError when trying to read the last few video frames. And video frames get from `get_data(index)` will always be frame of the wrong index

### ISSUE THREAD

**Issue by Nunatic02** [imageio/imageio#949]
# Issue Overview
For some videos, the ffmpeg reader created using `self.reader = imageio.get_reader(file_path, 'ffmpeg');` is able to get the right number of frames (`148` in this test case). But it is unable to get video frame data of the last few frames. e.g. `self.reader.get_data(147)`(the last frame) and `self.reader.get_data(146)`(the second to last frame) in our test will result in `IndexError`.

# Experiment Results
### Trial1: Open video in Adobe Premiere Pro, change timecode to be using frame number.
Get to know that frame number is from 0 to 147. So we have 148 frames.
### Trial 2: Use ffmpeg to extract all frames 
```
ffmpeg -i "input.mp4" "frames/%d.png"
```
We get all frames from `1.png` to `148.png`. Doubled checked that we have 148 frames.

### Trial 3: Try to read number of frames using `imageio` ffmpeg `reader`
```
self.reader = imageio.get_reader(file_path, 'ffmpeg'); 
if(hasattr(self.reader, 'count_frames')):
	reader_length = self.reader.count_frames();
else:
	reader_length = self.reader.get_length();
num_of_frames = int(reader_length)
```
result: `num_of_frames` is 148, which is the correct number of frames as checked in other methods above.

## Experiment 2: Trying to run the ffmpeg command used by `imageio-ffmpeg` directly 
After some debugging, I noticed that when calling `self.reader.get_data(147)`, `imageio-ffmpeg` is trying to run the ffmpeg command `ffmpeg -ss 0.000000 -i $VIDEO_FILE_PATH -pix_fmt rgb24 -vcodec rawvideo -f image2pipe -ss 5.055021 -`, which outputs
```
ffmpeg version 5.1.2 Copyright (c) 2000-2022 the FFmpeg developers
  built with Apple clang version 14.0.0 (clang-1400.0.29.202)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/5.1.2_4 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags= --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enable-libsoxr --enable-libzmq --enable-libzimg --disable-libjack --disable-indev=jack --enable-videotoolbox
  libavutil      57. 28.100 / 57. 28.100
  libavcodec     59. 37.100 / 59. 37.100
  libavformat    59. 27.100 / 59. 27.100
  libavdevice    59.  7.100 / 59.  7.100
  libavfilter     8. 44.100 /  8. 44.100
  libswscale      6.  7.100 /  6.  7.100
  libswresample   4.  7.100 /  4.  7.100
  libpostproc    56.  6.100 / 56.  6.100
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '/Users/nunatic/Dropbox/2023Spring/cs6682-SP2023/CS6682-SP2023-A1/part2/AbeVideos/AbeSelfIntroCrop_short.mp4':
  Metadata:
    major_brand     : mp42
    minor_version   : 512
    compatible_brands: mp42iso2avc1mp41
    creation_time   : 2023-02-13T23:30:06.000000Z
    encoder         : HandBrake 1.4.2 2021100300
  Duration: 00:00:05.04, start: 0.000000, bitrate: 1279 kb/s
  Stream #0:0[0x1](und): Video: h264 (Main) (avc1 / 0x31637661), yuv420p(tv, smpte170m/smpte170m/bt709, progressive), 620x680 [SAR 1:1 DAR 31:34], 1117 kb/s, 29.45 fps, 29.08 tbr, 90k tbn (default)
    Metadata:
      creation_time   : 2023-02-13T23:30:06.000000Z
      handler_name    : VideoHandler
      vendor_id       : [0][0][0][0]
  Stream #0:1[0x2](und): Audio: aac (LC) (mp4a / 0x6134706D), 44100 Hz, mono, fltp, 153 kb/s (default)
    Metadata:
      creation_time   : 2023-02-13T23:30:06.000000Z
      handler_name    : Mono
      vendor_id       : [0][0][0][0]
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> rawvideo (native))
Press [q] to stop, [?] for help
Output #0, image2pipe, to 'pipe:':
  Metadata:
    major_brand     : mp42
    minor_version   : 512
    compatible_brands: mp42iso2avc1mp41
    encoder         : Lavf59.27.100
  Stream #0:0(und): Video: rawvideo (RGB[24] / 0x18424752), rgb24, 620x680 [SAR 1:1 DAR 31:34], q=2-31, 294276 kb/s, 29.08 fps, 29.08 tbn (default)
    Metadata:
      creation_time   : 2023-02-13T23:30:06.000000Z
      handler_name    : VideoHandler
      vendor_id       : [0][0][0][0]
      encoder         : Lavc59.37.100 rawvideo
frame=    0 fps=0.0 q=0.0 Lsize=       0kB time=00:00:00.00 bitrate=N/A speed=   0x
video:0kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown
Output file is empty, nothing was encoded (check -ss / -t / -frames parameters if used)
```
Notice the last line mentions `Output file is empty, nothing was encoded (check -ss / -t / -frames parameters if used)`, meaning no video frame data has been outputed, because ffmpeg can't extract a video frame at that time.

ffmpeg code runned when calling `self.get_data(146)` is `ffmpeg -ss 0.000000 -i $VIDEO_FILE_PATH -pix_fmt rgb24 -vcodec rawvideo -f image2pipe -ss 5.020633 -,` , which also mentions `Output file is empty, nothing was encoded (check -ss / -t / -frames parameters if used)`

### Result Analysis
Output result of these 2 command makes sense, because the `fps` of the video is 29.45. So the starttime of the last 2 frames should be `147/29.45 = 4.9915110357` and `146/29.45=4.9575551783`. instead of `5.055021` and `5.020633`.

**Why they these 2 commands doesn't output video frame data?**
The time  `5.055021` is larger that the video duration, so there's no doubt that ffmpeg can't output any video frame data with that starttime.
The time `5.020633` is larger than starttime of the last frame (`4.9915110357`), so it also shouldn't output anything.
### Conclusion
**All of these point to the fact that starttime calculated to extract a certain frame is wrong.**

# Cause of Error
## TL;DR: 
 The error is caused by using a wrong `fps` metadata returned by `imageio-ffmpeg` .
 
 [`imageio-ffmpeg` is currently parsing `tbr` as frame rate metadata `fps`](https://github.com/imageio/imageio-ffmpeg/blob/77a3d85feeef431e2c8ad066eb7d1df08730337c/imageio_ffmpeg/_parsing.py#L160-L170) because of #262. However, we instead should still parse the `fps` value as frame rate from the ffmpeg banner instead of `tbr`. Because [the source code of `imageio` is treating `fps` as `average frame rate`, namely `total # of frames / total duration`](https://github.com/imageio/imageio/blob/b02aa05dea12331168ab413e56cb79c7e64b5b39/imageio/plugins/ffmpeg.py#L445), and in ffmpeg, value denoting average frame rate is `fps`, not `tbr`. (Proofs and references are as below). For special cases when `tbr` is preferred, [user should set it directly in the `fps` parameter of the ffmpeg reader](https://github.com/imageio/imageio/blob/b02aa05dea12331168ab413e56cb79c7e64b5b39/imageio/plugins/ffmpeg.py#L28-L33).

#### Some context knowledge
`index`: index of the frame we are trying to read, start from 0

##### Context 1: How `imageio` reads video frames

To read frames, `imageio` will [create a `imageio-ffmpeg`.`read_frames` generator and store it as `self._read_gen`](https://github.com/imageio/imageio/blob/b02aa05dea12331168ab413e56cb79c7e64b5b39/imageio/plugins/ffmpeg.py#L468-L470).  The `_read_gen` is a generator function, where calling [`self._read_gen().__next__()` each time will yield one video frame in byte data](https://github.com/imageio/imageio/blob/b02aa05dea12331168ab413e56cb79c7e64b5b39/imageio/plugins/ffmpeg.py#L514). (Special case: [when calling `__next()__` for the first time, `read_frames` function will return `meta`, which is a dictionary containing video's metadata like `fps`](https://github.com/imageio/imageio/blob/b02aa05dea12331168ab413e56cb79c7e64b5b39/imageio/plugins/ffmpeg.py#L476).

##### Context 2: How `imageio-ffmpeg` read byte data of video frames

`imageio-ffmpeg` will [spawn a subprocess](https://github.com/imageio/imageio-ffmpeg/blob/77a3d85feeef431e2c8ad066eb7d1df08730337c/imageio_ffmpeg/_io.py#L267-L273) [that ask `ffmpeg` to send video frame byte data over `pipe`](https://github.com/imageio/imageio-ffmpeg/blob/77a3d85feeef431e2c8ad066eb7d1df08730337c/imageio_ffmpeg/_io.py#L261). Because `imageio-ffmpeg` has [set `stdout` of the subprocess to be send to pipe](https://github.com/imageio/imageio-ffmpeg/blob/77a3d85feeef431e2c8ad066eb7d1df08730337c/imageio_ffmpeg/_io.py#L267-L273), we can then [read from `stdout` of the subprocess iteratively to get byte data of video frames](https://github.com/imageio/imageio-ffmpeg/blob/77a3d85feeef431e2c8ad066eb7d1df08730337c/imageio_ffmpeg/_io.py#L318-L329).

##### Context 3: How `imageio-ffmpeg`'s `read_frames` generator function read and yield metadata and video frames

The subprocess, spawned by the `imageio-ffmpeg`'s `read_frames` function, will run a command like `ffmpeg -ss $FAST_SEEK_TIME -i $PATH_TO_VIDEO_FILE -pix_fmt rgb24 -vcodec rawvideo -f image2pipe -ss $SLOW_SEEK_TIME -`
e.g. `ffmpeg -ss 0.000000 -i $PATH_TO_VIDEO_FILE -pix_fmt rgb24 -vcodec rawvideo -f image2pipe -ss 5.055021 -`

`-ss $FAST_SEEK_TIME` combined with `-ss $SLOW_SEEK_TIME` will seek in the stream foward by `$FAST_SEEK_TIME` + `$SLOW_SEEK_TIME`. The reason for seperating a fast seek time and a slow seek time is because setting `-ss $FAST_SEEK_TIME` before `-i` will seek faster, but not accurate. So it's like first fast forward rapidly by a large amount, and move frame by frame slowly by setting `-ss $SLOW_SEEK_TIME`.
(reference: 
[Seeking – FFmpeg](https://trac.ffmpeg.org/wiki/Seeking)
[imageio/ffmpeg.py at b02aa05dea12331168ab413e56cb79c7e64b5b39 · imageio/imageio · GitHub](https://github.com/imageio/imageio/blob/b02aa05dea12331168ab413e56cb79c7e64b5b39/imageio/plugins/ffmpeg.py#L437-L452))

With the command above, ffmpeg first seek the time of the video to `$FAST_SEEK_TIME+$SLOW_SEEK_TIME`, then start writing video frame data from that point in time to the pipe buffer. [`read_frames` function will read the pipe buffer until it fetch out enough data of one video frame size. It will then yield that data and pause.](https://github.com/imageio/imageio-ffmpeg/blob/77a3d85feeef431e2c8ad066eb7d1df08730337c/imageio_ffmpeg/_io.py#L318-L329)

##### Context 5: Where does the `fps` metadata used by `imageio`  comes from
For getting the `fps` metadata, [`imageio-ffmpeg` will try to parse the ffmpeg banner ouputed from the subprocess used above, collect metadata information (like `fps`) into a dictionary variable `meta`, and return it to `imageio`.  `imageio` will then use it to set its `self._meta`](https://github.com/imageio/imageio-ffmpeg/blob/77a3d85feeef431e2c8ad066eb7d1df08730337c/imageio_ffmpeg/_io.py#L301). 
**A problematic thing is that,[ it is perferring the `tbr` value of the ffmpeg banner as our `fps` metadata.**](https://github.com/imageio/imageio-ffmpeg/blob/77a3d85feeef431e2c8ad066eb7d1df08730337c/imageio_ffmpeg/_parsing.py#L165-L170) Because of this issue #262. I'll prove below that we shouldn't be using this `tbr` value as our `fps` metadata, because `imageio` is using `fps`  as `average frame rate`, while `tbr` is not `average frame rate`, but rather something called `r_frame_rate`. `fps` in the `ffmpeg` banner is the true `average frame rate`

Another issue is that, [the parsing code will actually parse the `tbr` value of the last output video source, instead of the input video source.](https://github.com/imageio/imageio-ffmpeg/blob/77a3d85feeef431e2c8ad066eb7d1df08730337c/imageio_ffmpeg/_parsing.py#L165-L170).Because the fps value will always be overwrite by data parsed from `videolines[-1]`.  And The output video's `fps` will be the same as the input video's `tbr`, so we should not use that one.
References: [How videolines are collected](https://github.com/imageio/imageio-ffmpeg/blob/77a3d85feeef431e2c8ad066eb7d1df08730337c/imageio_ffmpeg/_parsing.py#L131-L133)
Example ffmpeg banner:
```
ffmpeg version 5.1.2 Copyright (c) 2000-2022 the FFmpeg developers
  built with Apple clang version 14.0.0 (clang-1400.0.29.202)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/5.1.2_4 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags= --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enable-libsoxr --enable-libzmq --enable-libzimg --disable-libjack --disable-indev=jack --enable-videotoolbox
  libavutil      57. 28.100 / 57. 28.100
  libavcodec     59. 37.100 / 59. 37.100
  libavformat    59. 27.100 / 59. 27.100
  libavdevice    59.  7.100 / 59.  7.100
  libavfilter     8. 44.100 /  8. 44.100
  libswscale      6.  7.100 /  6.  7.100
  libswresample   4.  7.100 /  4.  7.100
  libpostproc    56.  6.100 / 56.  6.100
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '/Users/nunatic/Dropbox/2023Spring/cs6682-SP2023/CS6682-SP2023-A1/part2/AbeVideos/AbeSelfIntroCrop_short.mp4':
  Metadata:
    major_brand     : mp42
    minor_version   : 512
    compatible_brands: mp42iso2avc1mp41
    creation_time   : 2023-02-13T23:30:06.000000Z
    encoder         : HandBrake 1.4.2 2021100300
  Duration: 00:00:05.04, start: 0.000000, bitrate: 1279 kb/s
  Stream #0:0[0x1](und): Video: h264 (Main) (avc1 / 0x31637661), yuv420p(tv, smpte170m/smpte170m/bt709, progressive), 620x680 [SAR 1:1 DAR 31:34], 1117 kb/s, 29.45 fps, 29.08 tbr, 90k tbn (default)
    Metadata:
      creation_time   : 2023-02-13T23:30:06.000000Z
      handler_name    : VideoHandler
      vendor_id       : [0][0][0][0]
  Stream #0:1[0x2](und): Audio: aac (LC) (mp4a / 0x6134706D), 44100 Hz, mono, fltp, 153 kb/s (default)
    Metadata:
      creation_time   : 2023-02-13T23:30:06.000000Z
      handler_name    : Mono
      vendor_id       : [0][0][0][0]
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> rawvideo (native))
Press [q] to stop, [?] for help
Output #0, image2pipe, to 'pipe:':
  Metadata:
    major_brand     : mp42
    minor_version   : 512
    compatible_brands: mp42iso2avc1mp41
    encoder         : Lavf59.27.100
  Stream #0:0(und): Video: rawvideo (RGB[24] / 0x18424752), rgb24, 620x680 [SAR 1:1 DAR 31:34], q=2-31, 294276 kb/s, 29.08 fps, 29.08 tbn (default)
    Metadata:
      creation_time   : 2023-02-13T23:30:06.000000Z
      handler_name    : VideoHandler
      vendor_id       : [0][0][0][0]
      encoder         : Lavc59.37.100 rawvideo
```

##### Context 4: How `imageio`'s ffmpeg reader get frame data of a specific indexed frame
We can create a ffmpeg reader using the `self.reader = imageio.get_reader(file_path, 'ffmpeg')`
And we can get video frame data at a specific index using `self.reader.get_data(index)`, where index start from 0.
[When target `index` is smaller that the current video seeked position + 100 (reference), (where by default, the current video seeked position is -1), imageio will just skip through a lot of frames until it reaches the target `index` frame](https://github.com/imageio/imageio/blob/b02aa05dea12331168ab413e56cb79c7e64b5b39/imageio/plugins/ffmpeg.py#L407-L410). [However, when the target `index` is larger that the current video seeked frame + 100 (reference), it will be too expensive to skip a lot of frames. So `imageio` will use the combination of `-ss $FAST_SEEK_TIME` and `-ss $SLOW_SEEK_TIME`  as mentioned above, to ask ffmpeg to seek to the time of the target `index` frame.](https://github.com/imageio/imageio/blob/b02aa05dea12331168ab413e56cb79c7e64b5b39/imageio/plugins/ffmpeg.py#L437-L452)

How does `imageio` calculate the time of the target `index` frame? It calculated such time as [`starttime = index / self._meta["fps"]`](https://github.com/imageio/imageio/blob/b02aa05dea12331168ab413e56cb79c7e64b5b39/imageio/plugins/ffmpeg.py#L445) Here we can see that such calculation will only be correct, if `self._meta["fps"] * total # of frames = total duration`, namely we must have ``self._meta["fps"] = total # of frames / total duration` = `average frame rate`

##### Important takeaway: 
[the source code of `imageio` is treating `fps` as `average frame rate`, namely `total # of frames / total duration`](https://github.com/imageio/imageio/blob/b02aa05dea12331168ab413e56cb79c7e64b5b39/imageio/plugins/ffmpeg.py#L445)

### Why using a wrong `fps` value in the bug that the last few frames can't be read?
In our test case, the `tbr` value 29.08 is smaller than the `fps` value 29.45 (which the real average frame rate). Since we are using a `fps` smaller than reality, the calculated duration of a single frame will be larger than reality, and the starttime calculated for the last few frames will be larger than reality, and even exceed the video duration.

## Other impacts of the wrong `fps` issue

When `tbr` of the video is different from `fps`, `get_data(index)` will always return video frames of a wrong index, because of a wrong starttime calculated.

#### References
- [ffmpeg source code showing that `fps` is `avg_frame_rate`, `r_frame_rate`](https://github.com/FFmpeg/FFmpeg/blob/e5c11ab9cef60a6c9141d7a60145ea833dd88485/libavformat/dump.c#L518-L521)
- [whats the difference between r_frame_rate and avg_frame_rate ? · Issue #7 · eugeneware/ffprobe · GitHub](https://github.com/eugeneware/ffprobe/issues/7)
-  [What is fps, tbr, tbn, tbc in `ffmpeg`? - Super User](https://superuser.com/a/1759811/1189837)
-  [ffmpeg default output frame rate - Video Production Stack Exchange](https://video.stackexchange.com/questions/20789/ffmpeg-default-output-frame-rate?newreg=e797b27b58a241dc9af8734dc8e14dc4)

#### Experiment proof
In the ffmpeg output above, we can see that the duration of our test video is `00:00:05.04`, with `29.45 fps` and `29.08 tbr`. And we know that the # of total frames is 148. Using `fps` we get `29.45 * 5.04 = 148.428` while using `tbr` we get `29.08 * 5.04 = 146.5632`, so the `fps` value is a better number to represent average frame rate.

# In Summary
We should by default parse and use ffmpeg's `fps`, instead of `tbr`, as our `fps` metadata. Because the source code of `imageio` is using  `fps`  as `average frame rate`, namely `total # of frames / total duration`. And ffmpeg's `fps` value also denotes `average frame rate`, while `tbr` is just a guessed frame rate. If not doing so, `get_data(index)` will always return the wrong video frame data when `fps` and `tbr` of a video doesn't match. And even worse, trying run `get_data`  of the last few frames will result in fatal `IndexError`.

**Comment by Nunatic02** [comment]
I've created a [PR here](https://github.com/imageio/imageio-ffmpeg/pull/84) to fix the issue.

**Comment by almarklein** [comment]
This is the most complete issue report I've ever seen 🤯   Thanks a lot for all the digging, and providing all the background info!

**Comment by Nunatic02** [comment]
> This is the most complete issue report I've ever seen 🤯 Thanks a lot for all the digging, and providing all the background info!

Thank you so much Almar @almarklein!! Your comment is super encouraging!! 🥰🥰🥰🥰
And thank you so much for all the contributions you have made to imageio, imageio-ffmpeg and other projects!
